server/backend/tracking.py

```diff
@@ -0,0 +1,66 @@
+import linecache
+import os
+import tracemalloc
+
+from contextlib import contextmanager
+from timeit import default_timer
+
+def display_top(snapshot, key_type='lineno', limit=3):
+    snapshot = snapshot.filter_traces((
+        tracemalloc.Filter(False, "<frozen importlib._bootstrap>"),
+        tracemalloc.Filter(False, "<unknown>"),
+    ))
+    top_stats = snapshot.statistics(key_type)
+
+    print("Top %s lines" % limit)
+    for index, stat in enumerate(top_stats[:limit], 1):
+        frame = stat.traceback[0]
+        # replace "/path/to/module/file.py" with "module/file.py"
+        filename = os.sep.join(frame.filename.split(os.sep)[-2:])
+        print("#%s: %s:%s: %.1f KiB"
+              % (index, filename, frame.lineno, stat.size / 1024))
+        line = linecache.getline(frame.filename, frame.lineno).strip()
+        if line:
+            print('    %s' % line)
+
+    other = top_stats[limit:]
+    if other:
+        size = sum(stat.size for stat in other)
+        print("%s other: %.1f KiB" % (len(other), size / 1024))
+    total = sum(stat.size for stat in top_stats)
+    print("Total allocated size: %.1f KiB" % (total / 1024))
+
+
+@contextmanager
+def track_usage(limit=3):
+    tracemalloc.start()
+    yield
```

Comment:
I assume this yield is similar to wait? Not sure I following this approach.

Reply:
Sort of...the `@contextmanager` decorator creates a context manager-compatible class from a generator function (i.e., a function with a `yield` in it). A context manager is an interface that can be used with the `with` keyword; it has an `__enter__` method that's called before the `with` block is entered, and an `__exit__` method that's called once the block is exited (either normally or through an exception being thrown). Typically context managers are used to manage resources that need to be cleaned up, i.e. file pointers that need to be closed no matter how the code that's using them terminates. In this case I'm using a context manager as a convenience for wrapping some code for tracking its memory usage.

The `@contextmanager` decorator expects that there's at least one `yield` in the function (which makes it a generator), and basically does the following:
1. runs everything before the `yield` as the `__enter__` method,
2. uses the value that's yielded as the value of the context manager within the `with` block (in this case, None, since I didn't yield anything)
3. runs everything after the `yield` as the `__exit__` method.

In this case, I'm starting a memory usage tracker, then allowing the code within the `with` block (i.e., the code I'm monitoring for memory usage) to run, then collecting and displaying stats at the end.

Reply:
Ah! I gotcha. That is a pretty neat feature. Good to know!